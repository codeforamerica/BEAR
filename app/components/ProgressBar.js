import React, { Component } from 'react';

type Props = {
  fileSizeInBytes: number,
  onComplete: void => void
};

export const PROCESSING_RATE = 10000000;

export default class ProgressBar extends Component<Props> {
  constructor(props) {
    super(props);
    const { fileSizeInBytes } = this.props;
    const step = (PROCESSING_RATE * 100) / fileSizeInBytes;
    this.state = {
      fill: 0,
      step: Math.min(10, step)
    };
  }

  componentDidMount() {
    this.timerID = setInterval(() => this.tick(), 1000);
  }

  componentWillUnmount() {
    clearInterval(this.timerID);
  }

  tick() {
    this.setState(state => ({ fill: Math.min(100, state.fill + state.step) }));
    const { fill } = this.state;
    if (fill === 100) {
      clearInterval(this.timerID);
      const { onComplete } = this.props;
      onComplete();
    }
  }

  render() {
    const { fill } = this.state;
    return <div> Fill: {Math.round(fill)} </div>;
  }
}
