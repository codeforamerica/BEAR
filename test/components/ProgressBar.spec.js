import sinon from 'sinon';
import React from 'react';
import Enzyme, { shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import renderer from 'react-test-renderer';
import ProgressBar, {
  PROCESSING_RATE,
  MAX_STEP_SIZE
} from '../../app/components/ProgressBar';
import sleep from '../../app/utils/testHelpers';

Enzyme.configure({ adapter: new Adapter() });
const sandbox = sinon.createSandbox();

function setup(fileSize = 1000) {
  const onCompleteCallbackSpy = sandbox.spy();
  const component = shallow(
    <ProgressBar
      fileSizeInBytes={fileSize}
      onCompleteCallback={onCompleteCallbackSpy}
      isComplete={false}
    />
  );
  return {
    component,
    onCompleteCallbackSpy
  };
}

afterEach(() => {
  sandbox.restore();
});

describe('ProgressBar component', () => {
  describe('setup', () => {
    it('defaults the stepSize to the maximum if the computed stepSize would be > maximum', () => {
      const { component } = setup(PROCESSING_RATE / 100);
      expect(component.instance().state.stepSize).toEqual(MAX_STEP_SIZE);
    });

    it('uses the computed stepSize if the stepSize is <= maximum', () => {
      const { component } = setup(PROCESSING_RATE * 100);
      expect(component.instance().state.stepSize).toEqual(1);
    });

    it('initializes fill to zero', async () => {
      const { component } = setup();
      expect(component.instance().state.fill).toEqual(0);
    });
  });

  describe('tick', () => {
    it('increases fill by the stepSize amount', () => {
      const { component } = setup();
      const instance = component.instance();
      instance.tick();
      expect(instance.state.fill).toEqual(10);
      instance.tick();
      expect(instance.state.fill).toEqual(20);
    });
  });

  describe('lifecycle', () => {
    it('advances fill with the passage of time', async () => {
      const { component } = setup();
      expect(component.instance().state.fill).toEqual(0);
      await sleep(3);
      expect(component.instance().state.fill > 0).toBe(true);
    });
  });

  describe('termination', () => {
    describe('when fill is 100 but isComplete not set', () => {
      it('does not call the onCompleteCallback method', () => {
        const { component, onCompleteCallbackSpy } = setup();
        const instance = component.instance();

        for (let i = 0; i < 10; i += 1) {
          instance.tick();
        }
        expect(onCompleteCallbackSpy.called).toBe(false);
      });
    });

    describe('when fill is not 100 and isComplete is set', () => {
      it('does not call the onCompleteCallback method', () => {
        const { component, onCompleteCallbackSpy } = setup();
        const instance = component.instance();

        for (let i = 0; i < 3; i += 1) {
          instance.tick();
        }
        component.setProps({ isComplete: true });
        expect(onCompleteCallbackSpy.called).toBe(false);
      });
    });

    describe('when fill is 100 and isComplete is set', () => {
      it('calls the onCompleteCallback method', () => {
        const { component, onCompleteCallbackSpy } = setup();
        const instance = component.instance();

        component.setProps({ isComplete: true });
        instance.setState({ fill: 100 });
        instance.tick();
        expect(onCompleteCallbackSpy.called).toBe(true);
      });
    });
  });

  it('should match exact snapshot', () => {
    const { component } = setup();
    const tree = renderer.create(component).toJSON();

    expect(tree).toMatchSnapshot();
  });
});
