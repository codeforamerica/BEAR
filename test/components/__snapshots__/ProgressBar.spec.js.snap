// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProgressBar component should match exact snapshot 1`] = `
<div>
   Fill: 
  0
   
</div>
`;
